File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(sequence, predicate)
  for i = 1, #sequence do
    if predicate(sequence[i]) then
      return string.lower(sequence[i])
    end
  end
  return nil
end


-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.wrap(function()
    local power = 0
    local number = 1
    while number >= limit do
      coroutine.yield(value)
      power = power + 1
      number = base ^ power
    end
    coroutine.yield(nil)
  end)
end

-- Write your say function here
function say(message)
  local output = [' ']

  if message == nil do
    return output
  end

  function join(input)

  


  

-- Write your line count function here

-- Write your Quaternion table here

File: javascript/exercises.js
import { open } from "node:fs/promises"

export function change(amount) {
  if (!Number.isInteger(amount)) {
    throw new TypeError("Amount must be an integer")
  }
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let [counts, remaining] = [{}, amount]
  for (const denomination of [25, 10, 5, 1]) {
    counts[denomination] = Math.floor(remaining / denomination)
    remaining %= denomination
  }
  return counts
}

// Write your first then lower case function here
export function firstThenLowerCase(sequence, predicate) {
  for (let i = 0; i < sequence.length; i++) {
    if (predicate(sequence[i])) {   // does a satisfy p?
      return sequence[i]?.toLowerCase(); // if it does, make it lowercase and return a[i]. ?. returns undefined for unexpected cases
    }
  }
}

// Write your powers generator here
export function* powersGenerator({ofBase: base, upTo: limit}) { // define function as generator function w/ *
  const maxIterations = Math.floor(Math.log(limit) / Math.log(base)); // lil algebra to determine our limit
  let n = 0;

  while (n <= maxIterations) { // recursion to get our powahs
    const value = Math.pow(base, n);
    yield value; // use yield rather than return
    n += 1;
  }
}

// Write your say function here
export function say(word) {
  const message = []; // initialize output array

  function joiner(input) { // chain function 
    if (input === undefined) {
      return message.join(' '); // for no words/final case
    }
    message.push(input);
    return joiner;
  }
  
  if (word === undefined) {
    return message.join(' ');
  }

  message.push(word);
  return joiner;
}


// Write your line count function here

// const fs = require('fs');
// const readline = require('readline');

// export function meaningfulLineCount(file) {
//   new Promise((resolve, reject) => {

    
//   }

// }

// Write your Quaternion class here

class Quaternion {
  constructor (i, k, j) {
    this.i = i;
    this.j = j;
    this.k = k;
    this.x = x;
  }
  toString() {
    return '<$this.i'

}

}


